scripts/git_changes_utils.py

```diff
@@ -0,0 +1,306 @@
+    out, err = start_subprocess_for_result(git_cmd)
+    if not err:
+        file_list = []
+        for line in out.splitlines():
+            # The lines in the output generally look similar to these:
+            #
+            #   A\tfilename
+            #   M\tfilename
+            #   R100\toldfilename\tnewfilename
+            #
+            # We extract the first char (indicating the status), and the string
+            # after the last tab character.
+            file_list.append(
+                FileDiff(bytes([line[0]]), line[line.rfind(b'\t') + 1:]))
+        return file_list
+    else:
+        raise ValueError(err)
+
+
+def get_merge_base(branch: str, other_branch: str) -> str:
+    """Returns the most-recent commit shared by both branches. Order doesn't
+    matter.
+
+    The commit returned is the same one used on GitHub's UI for comparing pull
+    requests.
+
+    Args:
+        branch: str. A branch name or commit hash.
+        other_branch: str. A branch name or commit hash.
+
+    Returns:
+        str. The common commit hash shared by both branches.
+
+    Raises:
+        ValueError. An error occurred in the git command.
+    """
+    merge_base, err = start_subprocess_for_result(
+        ['git', 'merge-base', branch, other_branch])
+    if err:
+        raise ValueError(err)
+
+    return merge_base.decode('utf-8').strip()
+
+
+def compare_to_remote(
+    remote: str, local_branch: str, remote_branch: Optional[str] = None
+) -> List[FileDiff]:
+    """Compare local with remote branch with git diff.
+
+    Parameter:
+        remote: str. Name of the git remote being pushed to.
+        local_branch: str. Name of the git branch being pushed to.
+        remote_branch: str|None. The name of the branch on the remote
+            to test against. If None, the remote branch is considered
+            to be the same as the local branch.
+
+    Returns:
+        list(FileDiff). List of FileDiffs that are modified, changed,
+        renamed or added but not deleted.
+
+    Raises:
+        ValueError. Raise ValueError if git command fails.
+    """
+    remote_branch = remote_branch if remote_branch else local_branch
+    git_remote = '%s/%s' % (remote, remote_branch)
+    # Ensure that references to the remote branches exist on the local machine.
+    start_subprocess_for_result(['git', 'pull', remote])
+    # Only compare differences to the merge base of the local and remote
+    # branches (what GitHub shows in the files tab of pull requests).
+    return git_diff_name_status(
+        get_merge_base(git_remote, local_branch), local_branch)
+
+
+def get_parent_branch_name_for_diff() -> str:
+    """Returns remote branch name against which the diff has to be checked.
+
+    Returns:
+        str. The name of the remote branch.
+    """
+    if common.is_current_branch_a_hotfix_branch():
+        return 'release-%s' % common.get_current_release_version_number(
+            common.get_current_branch_name())
```

Comment:
Is this used anywhere?

Reply:
Don't know, though I suppose this is for hotfix branches; don't know if this is still relevant.